### PR TITLE
[Week2] Finish w2 challenges and fix w2 hw5 concat str problem

### PR DESCRIPTION
**Pull request by ishin4554**
**更新**
- 修正 w2 hw5 多一個 concat string 的問題
- 完成 w2 挑戰題跟超級挑戰題在 challenge 資料夾裡

**優化**
- 簡化 ASCII 比較的寫法

**心得**
- 本來以為已經完成的 w2 挑戰題，隔一天起床看看又看到 bug，果然 test case 很重要 XD
- w1 超級挑戰題還是沒有挑戰成功，因為挑戰了太久，已經來到了第二週，所以先推進度，但會擔心自己會不會就這樣慢慢不挑戰 w1 的這題 QQ 有點猶豫是不是應該用現成的工具 jq 先解決這個問題，還是忽視 edge case 的狀況 ......

### CODE REVIEW

File: homeworks/week2/challenge/maze.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  for (let i = 0; i < mazeArr.length - 1; i += 1) {
+    next = mazeArr[i + 1].indexOf('.', now);
+    if (next === -1) {
+      next = mazeArr[i + 1].indexOf('.', 0);
+    }
+    sum += Math.abs(now - next) + 1;
+    now = next;
+  }
+  return sum;
+}
```
Comment (aszx87410) [inline homeworks/week2/challenge/maze.js:15]: 這題沒有你想像中簡單XD 可以試試看這個範例：

``` js
console.log(maze([
  '#.#######',
  '#.......#',
  '#######.#',
  '##....#.#',
  '##.##.#.#',
  '##.##...#',
  '##.######',
  ]));
```

關鍵字：BFS

File: homeworks/week2/challenge/search.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// BinarySearch
+function search(arr, num) {
+  const numArr = arr;
+  let end = numArr.length - 1;
+  let begin = 0;
+  let half = Math.floor(begin + end / 2);
```
Comment (aszx87410) [inline homeworks/week2/challenge/search.js:6]: (begin + end) 建議刮起來

File: homeworks/week2/challenge/search.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        begin = half + 1;
+      } else {
+        end = half - 1;
+      }
+      half = Math.floor((begin + end) / 2);
+    } else {
+      return half;
+    }
+  }
+  return -1;
```
Comment (aszx87410) [inline homeworks/week2/challenge/search.js:19]: 上面這整段調一下順序可以變比較漂亮，而且上面 half 的計算也可以直接拉進來放在 while 裡面

``` js
while (begin <= end) {
  let half = Math.floor((begin + end) / 2);
  if (num === numArr[half]) {
    return half;
  }

  if (num > numArr[half]) {
    begin = half + 1;
  } else {
    end = half - 1;
  }
}
return -1;

```

File: homeworks/week2/hw2.js [hunk 4]
```diff
@@ -1,8 +1,8 @@
 // 給定一字串，把第一個字轉成大寫之後「回傳」，若第一個字不是英文字母則忽略。
 function capitalize(str) {
-  const firstASCII = str.charCodeAt(0);
   let newStr = str;
-  if (firstASCII >= 97 && firstASCII <= 122) {
+  // str[0] >= 'a' 就可以取到 ASCII Code
+  if (str[0] >= 'a' && str[0] <= 'z') {
```
Comment (aszx87410) [inline homeworks/week2/hw2.js:5]: 仔細想想，你其實會發現這一行拿掉也可以XD

**Comment by aszx87410** [comment]
超級挑戰題把一些 edge case 考慮進去之後真的滿麻煩，我當初出題時也沒有想那麼細XD

考驗如何用內建的指令做複雜的字串處理不是我的原意（我沒有預期到會這麼複雜哈哈），建議可以交有 bug 的版本就好，以後有空再考慮優化

File: homeworks/week2/challenge/maze.js (hunk 1 above)
Comment (ishin4554) [reply homeworks/week2/challenge/maze.js:15]: 居然還有這種 QQ ！！！！ 果然 test case 也是一門學問
